src/mpid/ch4/netmod/include/netmod.h

```diff
@@ -63,6 +63,7 @@ typedef int (*MPIDI_NM_mpi_win_create_dynamic_hook_t) (MPIR_Win * win);
 typedef int (*MPIDI_NM_mpi_win_attach_hook_t) (MPIR_Win * win, void *base, MPI_Aint size);
 typedef int (*MPIDI_NM_mpi_win_detach_hook_t) (MPIR_Win * win, const void *base);
 typedef int (*MPIDI_NM_mpi_win_free_hook_t) (MPIR_Win * win);
+typedef int (*MPIDI_NM_mpi_win_set_info_hook_t) (MPIR_Win * win);
```

Comment:
1. >With this approach, it may become tricky for shmmod/netmod to know what has changed since last time

Right, I do not have a better solution to only notify the "updates". So the netmod/shmmod has to reload all hints that they internally use (e.g., `MPIDI_OFI_load_acc_hint`). Since this is not a perf-critical path, I think it is fine.

I do not want to "reload hints" at op issuing time, the "reloading" overhead (e.g., compute `dtypes_max_count` for all supported ops) is expensive.

2. >Is netmod/shmmod allowed to modify the info hint values?

I just checked the [mpi-standard master branch](https://github.com/mpi-forum/mpi-standard/blob/mpi-3.x/chap-one-side/one-side-2.tex), it says
```
\mpifunc{MPI\_WIN\_GET\_INFO} returns a new info object containing
the hints of the window associated with \mpiarg{win}.
The current setting of all hints related to this window is returned in
\mpiarg{info\_used}.  An \MPI/ implementation is required to return all hints
that are supported by the implementation and have default values
specified; any user-supplied hints that were not ignored by the
implementation; and any additional hints that were set by the
implementation.
```

So, we have to return the actual value of all supported hints in future MPI standard.

Reply:
To notify the "updates", one possible option would be to pass the `MPIR_Info` object to this hook. But that way we will have to re-parse the string values from the info object, in contrast to how you are parsing them from a single location now. Performance wise, maybe we can afford it, as it is not a performance critical path? THere can be arguments from software-engineering perspective though.

`get_info` is the also way for users to learn about the underlying hint values in case of communicator info hints.  (including default and rejected values)

Reply:
I do not like the way where netmod directly reads user hints (e.g., deserialize string). And I think `get_info` output should be controlled by CH4 as it manages underlying hints both from shmmod and netmod.

Reply:
Maybe we can discuss a better hint interface design in a separate PR. This PR wants to simply focus on new atomic hints and expose them to the netmod.

Reply:
I agree that directly reading the user hints at netmod may not be a very clean way.

Sure, we can discuss more about the hints interface in a separate PR. Adding some relevant thoughts here before I forget them: 

- I think the current model (reading at CH4) works fine if we assume CH4 is aware of all the hints defined at every netmod and shmmod implementations.
- But it might get little tricky if we were to allow lets say netmod to add new hints that CH4 is not aware of. I'm not sure if this is a valid scenario for RMA window hints. If it is valid, then we may get benefits by allowing netmod (and shmmod) layer to  process the hints in some way.

Reply:
After thought about the info updates again, I think it is not safe to simply update OFI level hints info (i.e., `MPIDI_OFI_WIN(win).acc_hint`) at `win_set_info`. Because this update may change the atomic operation's issuing routine (e.g., am -> hw). We need additional `am+hw flush` to ensure all previous atomics are completed. Thus, I think it might be simpler if we only allow user to set the new hints at window creation.

@hajimefu If you agree, I'd like to delete the `win_set_info_hook` commit. This could also help @sagarth to redesign info interface in a separate PR.

Reply:
@minsii I'm fine with that, but does it affect any functionality of this PR?

Reply:
I think it is fine for this PR. The user can still pass atomics hints at window creation/allocation time, which I believe is the most common use case. 

It is incorrect to update those hints at win_set_info, unless we find a good way to transparently ensure all pes in the window have flushed its outstanding atomics which were relying on previous hint value.

Both 3.1 standard and the master version allow the implementation to ignore hint updates. 
`
Advice to users. Some info items that an implementation can use when it creates a window cannot easily be changed once the window has been created. Thus, an implementation may ignore hints issued in this call that it would have accepted in a creation call.
`

Reply:
> Both 3.1 standard and the master version allow the implementation to ignore hint updates.

Exactly.

My another question here was what happens in the following case:
1. user gives some hints (say, "keyA=valX")
2. MPI decided to ignore it and defaulted to a more conservative value "valY".
3. user calls `MPI_Win_get_info`. What should MPI return to the user as the value of "keyA"?

Reply:
I think it should return "valY". It is unclear in current 3.1 standard, but it should be clarified in standard master branch.

Reply:
Just deleted `win_set_info_hook`, and added comment/code in `MPIDI_CH4I_win_set_info` to limit the update of atomic hints.

Reply:
Looks like the MPI standard 3.x document says something about `MPI_Win_get_info` for the ignored values
https://github.com/mpi-forum/mpi-standard/blob/mpi-3.x/chap-one-side/one-side-2.tex#L977
```
An implementation is free to ignore all hints; however, applications must comply 
with any info hints they provide that are used by the MPI implementation 
(i.e., are returned by a call to MPI_WIN_GET_INFO) and that place a restriction 
on the behavior of the application. 
```

Reply:
Yes, we should return the actual value internally being used.